### PR TITLE
TropicalGeometry: fixed det over tropical polynomials

### LINKED ISSUE #3714: Calculating `minors` of a matrix containing tropical polynomials fails
**Describe the bug**
When constructing a matrix `A` with tropical polynomials as entries and attempting to calculate its minors with `minors(A,n)`,
it fails because the computation attempts to use subtraction.

**To Reproduce**
```
using Oscar
T = tropical_semiring()
R, (x,y) = T[:x,:y]

A = identity_matrix(R,2)
minors(A,1)
```

**Expected behavior**
`minors(A,1)` should not fail and subtraction in the calculation of minors should be handled appropriately for tropical polynomials.

**System (please complete the following information):**
```
julia> Oscar.versioninfo(full=true)
OSCAR version 1.0.2
  combining:
    AbstractAlgebra.jl   v0.40.9
    GAP.jl               v0.10.3
    Hecke.jl             v0.30.9
    Nemo.jl              v0.43.2
    Polymake.jl          v0.11.15
    Singular.jl          v0.22.6
  building on:
    Antic_jll               v0.201.500+0
    Arb_jll                 v200.2300.0+0
    Calcium_jll             v0.401.100+0
    FLINT_jll               v200.900.9+0
    GAP_jll                 v400.1200.200+9
    Singular_jll            v403.216.1602+0
    libpolymake_julia_jll   v0.11.4+0
    libsingular_julia_jll   v0.44.2+0
    polymake_jll            v400.1100.1+0
See `]st -m` for a full list of dependencies.

Julia Version 1.10.2
Commit bd47eca2c8a (2024-03-01 10:14 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: macOS (x86_64-apple-darwin22.4.0)
  CPU: 4 × Intel(R) Core(TM) i5-7Y54 CPU @ 1.20GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-15.0.7 (ORCJIT, skylake)
Threads: 1 default, 0 interactive, 1 GC (on 4 virtual cores)
Environment:
  LD_LIBRARY_PATH = /Applications/Macaulay2-1.21/lib
Official https://julialang.org/ release
```

**Additional context**
I've included the stack trace for `minors(A,1)` in above example for convenience.
```
ERROR: Tropical subtraction not defined (use tropical division for classical subtraction)
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:35
 [2] -(::TropicalSemiringElem{typeof(min)})
   @ Oscar ~/.julia/packages/Oscar/CziNR/src/TropicalGeometry/semiring.jl:342
 [3] -(a::AbstractAlgebra.Generic.MPoly{TropicalSemiringElem{typeof(min)}})
   @ AbstractAlgebra.Generic ~/.julia/packages/AbstractAlgebra/wVuNz/src/generic/MPoly.jl:1015
 [4] charpoly(S::AbstractAlgebra.Generic.PolyRing{…}, Y::AbstractAlgebra.Generic.MatSpaceElem{…})
   @ AbstractAlgebra ~/.julia/packages/AbstractAlgebra/wVuNz/src/Matrix.jl:3995
 [5] det_df(M::AbstractAlgebra.Generic.MatSpaceElem{AbstractAlgebra.Generic.MPoly{TropicalSemiringElem{typeof(min)}}})
   @ AbstractAlgebra ~/.julia/packages/AbstractAlgebra/wVuNz/src/Matrix.jl:2259
 [6] det(M::AbstractAlgebra.Generic.MatSpaceElem{AbstractAlgebra.Generic.MPoly{TropicalSemiringElem{typeof(min)}}})
   @ AbstractAlgebra ~/.julia/packages/AbstractAlgebra/wVuNz/src/Matrix.jl:2308
 [7] minors(A::AbstractAlgebra.Generic.MatSpaceElem{AbstractAlgebra.Generic.MPoly{TropicalSemiringElem{…}}}, k::Int64)
   @ AbstractAlgebra ~/.julia/packages/AbstractAlgebra/wVuNz/src/Matrix.jl:2429
 [8] top-level scope
   @ REPL[15]:1

caused by: Tropical subtraction not defined (use tropical division for classical subtraction)
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:35
 [2] -(::TropicalSemiringElem{typeof(min)})
   @ Oscar ~/.julia/packages/Oscar/CziNR/src/TropicalGeometry/semiring.jl:342
 [3] -(a::AbstractAlgebra.Generic.MPoly{TropicalSemiringElem{typeof(min)}})
   @ AbstractAlgebra.Generic ~/.julia/packages/AbstractAlgebra/wVuNz/src/generic/MPoly.jl:1015
 [4] fflu!(P::Perm{Int64}, A::AbstractAlgebra.Generic.MatSpaceElem{AbstractAlgebra.Generic.MPoly{TropicalSemiringElem{…}}})
   @ AbstractAlgebra ~/.julia/packages/AbstractAlgebra/wVuNz/src/Matrix.jl:1760
 [5] det_fflu(M::AbstractAlgebra.Generic.MatSpaceElem{AbstractAlgebra.Generic.MPoly{TropicalSemiringElem{typeof(min)}}})
   @ AbstractAlgebra ~/.julia/packages/AbstractAlgebra/wVuNz/src/Matrix.jl:2271
 [6] det(M::AbstractAlgebra.Generic.MatSpaceElem{AbstractAlgebra.Generic.MPoly{TropicalSemiringElem{typeof(min)}}})
   @ AbstractAlgebra ~/.julia/packages/AbstractAlgebra/wVuNz/src/Matrix.jl:2306
 [7] minors(A::AbstractAlgebra.Generic.MatSpaceElem{AbstractAlgebra.Generic.MPoly{TropicalSemiringElem{…}}}, k::Int64)
   @ AbstractAlgebra ~/.julia/packages/AbstractAlgebra/wVuNz/src/Matrix.jl:2429
 [8] top-level scope
   @ REPL[15]:1
Some type information was truncated. Use `show(err)` to see complete types.
```

### PR DESCRIPTION
**Pull request by YueRen**
Fixes https://github.com/oscar-system/Oscar.jl/issues/3714.  Is there any better way than writing
```
function det(A::Generic.MatSpaceElem{R}) where {R<:Union{TropicalSemiringElem,MPolyRingElem{<:TropicalSemiringElem},PolyRingElem{<:TropicalSemiringElem}}}
```

### CODE REVIEW

**Comment by YueRen** [comment]
I also tried adding @ooinaruhugh as reviewer, but somehow that was not possible.

**Comment by thofma** [comment]
No, there is no better way.

**Review by thofma — APPROVED** [review]

File: src/TropicalGeometry/matrix.jl [hunk 1]
```diff
@@ -38,6 +38,6 @@ function det(A::Generic.MatSpaceElem{<:TropicalSemiringElem})
     return detA
 end
 
-function det(A::Matrix{TropicalSemiringElem{minOrMax}}) where {minOrMax<:Union{typeof(min),typeof(max)}}
+function det(A::Matrix{R}) where {R<:Union{TropicalSemiringElem,MPolyRingElem{<:TropicalSemiringElem},PolyRingElem{<:TropicalSemiringElem}}}
     return det(matrix(TropicalSemiring{minOrMax},A))
```
Comment (lgoettgens) [inline src/TropicalGeometry/matrix.jl]: This coercion won't work for polys and mpolys.
Could you add a test with a julia matrix of tropical polys exercising this?

Comment (YueRen) [reply src/TropicalGeometry/matrix.jl]: True and good catch.  I've tried to fix it, let's see whether it passes the tests.

**Comment by ooinaruhugh** [comment]
> I also tried adding @ooinaruhugh as reviewer, but somehow that was not possible.

At least I looked into the code and also tried using `minors`, my original works now. Thanks!

**Comment by YueRen** [comment]
The tests seem to pass, if somebody could (squash) merge it I'd be much obliged.

**Pull request merged**